blockchain/blockchain.go

```diff
@@ -0,0 +1,182 @@
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	JobPendingState    = "PENDING"
+	JobFundedState     = "FUNDED"
+	JobAddressHeader   = "snet-job-address"
+	JobSignatureHeader = "snet-job-signature"
+)
+
+type jobInfo struct {
+	jobAddressBytes   []byte
+	jobSignatureBytes []byte
+}
+
+var (
+	ethClient          *ethclient.Client
+	rawClient          *rpc.Client
+	agent              *Agent
+	privateKey         *ecdsa.PrivateKey
+	address            string
+	jobCompletionQueue chan *jobInfo
+)
+
+func init() {
+	if config.GetBool(config.BlockchainEnabledKey) {
+		// Setup ethereum client
+		if client, err := rpc.Dial(config.GetString(config.EthereumJsonRpcEndpointKey)); err != nil {
+			log.WithError(err).Panic("error creating rpc client")
+		} else {
+			rawClient = client
+			ethClient = ethclient.NewClient(rawClient)
+		}
+
+		// Setup agent
+		if a, err := NewAgent(common.HexToAddress(config.GetString(config.AgentContractAddressKey)), ethClient); err != nil {
+			log.WithError(err).Panic("error instantiating agent")
+		} else {
+			agent = a
+		}
+
+		// Setup identity
+		if privateKeyString := config.GetString(config.PrivateKeyKey); privateKeyString != "" {
+			if privKey, err := crypto.HexToECDSA(privateKeyString); err != nil {
+				log.WithError(err).Panic("error getting private key")
+			} else {
+				privateKey = privKey
+				address = crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+			}
+		} else if hdwalletMnemonic := config.GetString(config.HdwalletMnemonicKey); hdwalletMnemonic != "" {
+			if privKey, err := derivePrivateKey(hdwalletMnemonic, 44, 60, 0, 0, uint32(config.GetInt(config.HdwalletIndexKey))); err != nil {
+				log.WithError(err).Panic("error deriving private key")
+			} else {
+				privateKey = privKey
+				address = crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+			}
+		}
+
+		// Start event and job completion routines
+		jobCompletionQueue = make(chan *jobInfo, 1000)
+		go processJobCompletions()
+		go processEvents()
+		go submitOldJobsForCompletion()
+	}
+}
+
+func GetGrpcStreamInterceptor() grpc.StreamServerInterceptor {
+	if config.GetBool(config.BlockchainEnabledKey) {
+		return jobValidationInterceptor
+	} else {
+		return noOpInterceptor
+	}
+}
+
+func IsValidJobInvocation(jobAddressBytes, jobSignatureBytes []byte) bool {
+	log := log.WithFields(log.Fields{
+		"jobAddress":   common.BytesToAddress(jobAddressBytes).Hex(),
+		"jobSignature": hex.EncodeToString(jobSignatureBytes)})
+
+	// Get job state from db
+	log.Debug("retrieving job from database")
+	job := &db.Job{}
+
+	db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(db.JobBucketName)
+		jobBytes := bucket.Get(jobAddressBytes)
+		if jobBytes != nil {
+			json.Unmarshal(jobBytes, job)
+		}
+		return nil
+	})
+
+	// If job is marked completed locally, reject
+	if job.Completed {
+		log.Debug("job already marked completed locally")
+		return false
+	}
+
+	v, r, s, err := parseSignature(jobSignatureBytes)
```

Comment:
Four is probably too many things to be returning from a function. Suggest making it a struct with fields/methods on it. IOW, `(typ, error)` should be the return signature.